### PR TITLE
Form Function

### PR DESCRIPTION
**Pull request by sb3958**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "python.pythonPath": "c:\\Users\\Sharelle\\Desktop\\Senior Design\\conteo\\venv\\Scripts\\python.exe"
```
Comment (mneedleman) [inline .vscode/settings.json]: `.vscode` should probably be in the `.gitignore`. We don't need this in the repo

**Review by mneedleman — CHANGES_REQUESTED** [review]
We should discuss whether these pages should be authenticated routes or not.

File: api/urls.py [hunk 2]
```diff
@@ -21,4 +21,6 @@
     path("reports/", views.Reports.as_view()),
     path("accounts/", views.Accounts.as_view()),
     path("user/", views.UserAuthView.as_view()),
+    path("contact/", views.ContactUsView.as_view()),
+    path("feedback/", views.FeedbackView.as_view()),
```
Comment (elp49) [inline api/urls.py]: these views do not exist

File: api/serializers.py [hunk 3]
```diff
@@ -205,3 +214,15 @@ def to_representation(self, instance):
...
+class ContactUsSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ContactUs
+        fields = ("email", "name", "message")
+
+
+class FeedbackSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Feedback
+        fields = ("reason", "message", "email")
```
Comment (elp49) [inline api/serializers.py]: I think these fields can be `fields = '__all__'`

File: frontend/src/FullPageRoutes/AccountsPage.tsx [hunk 4]
```diff
@@ -69,7 +71,8 @@ export default function AccountsPage(): JSX.Element {
     editModeOn(!editMode)
   }
 
-  const handleSave = () => {
+  const handleSave = (e: any) => {
+    e.preventDefault().then(history.push("/"))
```
Comment (elp49) [inline frontend/src/FullPageRoutes/AccountsPage.tsx]: preventDefault doesn't return a value. do we really want to redirect after a user saves?

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 5]
```diff
@@ -50,13 +52,23 @@ const contactValue = {
...
     const { name, value } = e.target
     setValues({
       ...values,
       [name]: value,
     })
+    return request({
+      path: "/api/contact/",
+      method: "post",
+      body: contactValue,
+    })
```
Comment (elp49) [inline frontend/src/FullPageRoutes/ContactUs.tsx:68]: the request should be in `handleSubmit`

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 6]
```diff
@@ -50,13 +52,23 @@ const contactValue = {
 export default function ContactUs(): JSX.Element {
   const [values, setValues] = useState(contactValue)
   const classes = useStyles()
+  const history = useHistory()
+
+  const handleSubmit = (e: any) => {
+    e.preventDefault().then(history.push("/help"))
```
Comment (elp49) [inline frontend/src/FullPageRoutes/ContactUs.tsx]: preventDefault doesn't return a value.

File: frontend/src/FullPageRoutes/FeedbackPage.tsx [hunk 7]
```diff
@@ -39,13 +41,23 @@ const feedBackValue = {
 export default function FeedbackPage(): JSX.Element {
   const [values, setValues] = useState(feedBackValue)
   const classes = useStyles()
+  const history = useHistory()
+
+  const handleSubmit = (e: any) => {
+    e.preventDefault().then(history.push("/help"))
```
Comment (elp49) [inline frontend/src/FullPageRoutes/FeedbackPage.tsx]: preventDefault doesn't return a value.

File: frontend/src/FullPageRoutes/FeedbackPage.tsx [hunk 8]
```diff
@@ -39,13 +41,23 @@ const feedBackValue = {
...
     const { name, value } = e.target
     setValues({
       ...values,
       [name]: value,
     })
+    return request({
+      path: "/api/feedback/",
+      method: "post",
+      body: feedBackValue,
+    })
```
Comment (elp49) [inline frontend/src/FullPageRoutes/FeedbackPage.tsx]: the request should be in `handleSubmit`

**Review by elp49 — CHANGES_REQUESTED** [review]

File: .vscode/settings.json (hunk 1 above)
Comment (mneedleman) [reply .vscode/settings.json]: This still needs to be removed from the pushed files

File: api/models.py [hunk 9]
```diff
@@ -140,3 +140,17 @@ class ReportType(models.TextChoices):
...
+    name = models.CharField(max_length=50, blank=True)
+    message = models.CharField(max_length=2000, null=True)
+
+
+class Feedback(models.Model):
+    FEEDBACK_REASON = (("W", "Website"), ("Pp", "Penpal"), ("P", "Performance"))
+
+    reason = models.CharField(max_length=2, choices=FEEDBACK_REASON, null=True)
+    message = models.CharField(max_length=2000, blank=True)
+    email = models.CharField(max_length=50, blank=True)
```
Comment (mneedleman) [inline api/models.py]: I think there needs to be a generated migration for this to work properly

File: api/views.py [hunk 10]
```diff
@@ -416,3 +418,30 @@ class Accounts(generics.RetrieveAPIView):
...
+    serializer_class = ContactUsSerializer
+
+    def post(self, request):
+        email = request.data["email"]
+        name = request.data["name"]
+        message = request.data["message"]
+        serializer = self.serializer_class(
+            context={"email": email, "name": name, "message": message}
+        )
+        return response.Response(serializer.data, status=status.HTTP_201_CREATED)
```
Comment (mneedleman) [inline api/views.py]: I don't think you need to pull those out of the request and pass them in with the context. You should be able to pass the request

File: api/views.py [hunk 11]
```diff
@@ -416,3 +418,30 @@ class Accounts(generics.RetrieveAPIView):
...
+
+    def post(self, request):
+        reason = response.data["reason"]
+        message = response.data["message"]
+        email = response.data["email"]
+        if User.objects.get(request.user):
+            serializer = self.serializer_class(
+                context={"reason": reason, "message": message, "email": email}
+            )
+            return response.Response(serializer.data, status=status.HTTP_201_CREATED)
```
Comment (mneedleman) [inline api/views.py]: I'm not sure the check for user is necessary

File: api/views.py [hunk 12]
```diff
@@ -416,3 +418,30 @@ class Accounts(generics.RetrieveAPIView):
     def get(self, request):
         serializer = AccountSerializer(request.user)
         return response.Response(serializer.data, status=status.HTTP_200_OK)
+
+
+class ContactUs(generics.CreateAPIView):
+    serializer_class = ContactUsSerializer
```
Comment (mneedleman) [inline api/views.py:453]: Since the ContactUs page does not require the user to be logged in, we should add the AllowAny permission
```suggestion
class ContactUs(generics.CreateAPIView):
    serializer_class = ContactUsSerializer
    permission_clsses = (permissions.AllowAny)
```

File: frontend/src/FullPageRoutes/AccountsPage.tsx (hunk 4 above)
Comment (mneedleman) [reply frontend/src/FullPageRoutes/AccountsPage.tsx]: I agree with Eddie

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 13]
```diff
@@ -51,12 +52,17 @@ export default function ContactUs(): JSX.Element {
...
     const { name, value } = e.target
     setValues({
       ...values,
       [name]: value,
     })
+    return request({
+      path: "/api/contact/",
+      method: "post",
+      body: contactValue,
+    }).then(e.preventDefault())
```
Comment (mneedleman) [inline frontend/src/FullPageRoutes/ContactUs.tsx]: Typically `e.preventDefault()` is used to ensure that the click event does not propagate. What are you trying to do here?

**Review by mneedleman — CHANGES_REQUESTED** [review]

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 14]
```diff
@@ -51,12 +52,17 @@ export default function ContactUs(): JSX.Element {
   const [values, setValues] = useState(contactValue)
   const classes = useStyles()
 
-  const setContactValue = (e: any) => {
```
Comment (elp49) [inline frontend/src/FullPageRoutes/ContactUs.tsx]: why did you get rid of the set function? We will need both setContactValue and handleSubmit

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 15]
```diff
@@ -99,14 +105,14 @@ export default function ContactUs(): JSX.Element {
                   label="Your name"
                   name="name"
                   value={values.name}
-                  onChange={setContactValue}
+                  onChange={handleSubmit}
```
Comment (elp49) [inline frontend/src/FullPageRoutes/ContactUs.tsx]: why did you change this to handleSubmit?

File: frontend/src/FullPageRoutes/FeedbackPage.tsx [hunk 16]
```diff
@@ -74,7 +80,7 @@ export default function FeedbackPage(): JSX.Element {
                   id="reasons"
                   name="reasons"
                   value={values.reason}
-                  onChange={setFeedbackValue}
+                  onChange={handleSubmit}
```
Comment (elp49) [inline frontend/src/FullPageRoutes/FeedbackPage.tsx]: this should be setFeedbackValue

File: frontend/src/FullPageRoutes/FeedbackPage.tsx [hunk 17]
```diff
@@ -91,14 +97,14 @@ export default function FeedbackPage(): JSX.Element {
                   multiline
                   rowsMax={6}
                   value={values.message}
-                  onChange={setFeedbackValue}
+                  onChange={handleSubmit}
```
Comment (elp49) [inline frontend/src/FullPageRoutes/FeedbackPage.tsx]: this should be setFeedbackValue

File: frontend/src/FullPageRoutes/FeedbackPage.tsx [hunk 18]
```diff
@@ -91,14 +97,14 @@ export default function FeedbackPage(): JSX.Element {
...
                 />
                 <TextField
                   variant="outlined"
                   label="Your email"
                   name="email"
                   value={values.email}
-                  onChange={setFeedbackValue}
+                  onChange={handleSubmit}
```
Comment (elp49) [inline frontend/src/FullPageRoutes/FeedbackPage.tsx]: this should be setFeedbackValue

File: frontend/src/FullPageRoutes/FeedbackPage.tsx [hunk 19]
```diff
@@ -40,12 +41,17 @@ export default function FeedbackPage(): JSX.Element {
   const [values, setValues] = useState(feedBackValue)
   const classes = useStyles()
 
-  const setFeedbackValue = (e: any) => {
+  const handleSubmit = (e: any) => {
```
Comment (elp49) [inline frontend/src/FullPageRoutes/FeedbackPage.tsx]: we will also need setFeedbackValue

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 20]
```diff
@@ -99,14 +105,14 @@ export default function ContactUs(): JSX.Element {
...
                 />
                 <TextField
                   variant="outlined"
                   label="Your email"
                   name="email"
                   value={values.email}
-                  onChange={setContactValue}
+                  onChange={handleSubmit}
```
Comment (elp49) [inline frontend/src/FullPageRoutes/ContactUs.tsx]: this should be setContactValue

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 21]
```diff
@@ -115,7 +121,7 @@ export default function ContactUs(): JSX.Element {
                   multiline
                   rowsMax={6}
                   value={values.message}
-                  onChange={setContactValue}
+                  onChange={handleSubmit}
```
Comment (elp49) [inline frontend/src/FullPageRoutes/ContactUs.tsx]: this should be setContactValue

File: frontend/src/FullPageRoutes/ContactUs.tsx [hunk 22]
```diff
@@ -125,7 +131,7 @@ export default function ContactUs(): JSX.Element {
                   className={classes.submitButton}
                   size="large"
                   type="submit"
-                  href="#"
+                  onSubmit={handleSubmit}
```
Comment (elp49) [inline frontend/src/FullPageRoutes/ContactUs.tsx:137]: good

**Review by elp49 — CHANGES_REQUESTED** [review]

File: frontend/src/FullPageRoutes/ContactUs.tsx (hunk 14 above)
Comment (sb3958) [reply frontend/src/FullPageRoutes/ContactUs.tsx]: So just put the return request in the handleSubmit.

Comment (elp49) [reply frontend/src/FullPageRoutes/ContactUs.tsx]: right, you don't want to make the request until the user selects the submit button, but setContactValue should be called in the onChange of each input

File: api/migrations/0011_contactus_feedback.py [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+                        choices=[
+                            ("W", "Website"),
+                            ("Pp", "Penpal"),
+                            ("P", "Performance"),
+                        ],
+                        max_length=2,
+                        null=True,
+                    ),
+                ),
+                ("message", models.CharField(blank=True, max_length=2000)),
```
Comment (AyalaCnaan) [inline api/migrations/0011_contactus_feedback.py:54]: Just curious why this limit on message length.

File: frontend/src/FullPageRoutes/AccountsPage.tsx (hunk 4 above)
Comment (AyalaCnaan) [reply frontend/src/FullPageRoutes/AccountsPage.tsx]: Just curious, but what does it do?

File: api/migrations/0011_contactus_feedback.py (hunk 23 above)
Comment (sb3958) [reply api/migrations/0011_contactus_feedback.py:54]: @AyalaCnaan When a user sends their feedback, the maximum length the database will hold is 2000.

Comment (jcarrete5) [reply api/migrations/0011_contactus_feedback.py:54]: For technical reasons, [the `CharField` requires a max length be specified](https://docs.djangoproject.com/en/3.2/ref/models/fields/#django.db.models.CharField.max_length). A [`TextField`](https://docs.djangoproject.com/en/3.2/ref/models/fields/#textfield) does not have this restriction and would probably be better suited for something like the feedback message.

As for why 2000 was specifically chosen... :man_shrugging:

File: api/models.py (hunk 9 above)
Comment (jcarrete5) [reply api/models.py]: Looks like the `0011_contactus_feedback.py` creates this model.

File: api/models.py [hunk 24]
```diff
@@ -140,3 +140,17 @@ class ReportType(models.TextChoices):
...
+    name = models.CharField(max_length=50, blank=True)
+    message = models.CharField(max_length=2000, null=True)
+
+
+class Feedback(models.Model):
+    FEEDBACK_REASON = (("W", "Website"), ("Pp", "Penpal"), ("P", "Performance"))
+
+    reason = models.CharField(max_length=2, choices=FEEDBACK_REASON, null=True)
+    message = models.CharField(max_length=2000, blank=True)
```
Comment (jcarrete5) [inline api/models.py]: I think this would be better suited as a [TextField](https://docs.djangoproject.com/en/3.2/ref/models/fields/#textfield) rather than a CharField since the message could be arbitrarily large.

File: api/views.py (hunk 12 above)
Comment (jcarrete5) [reply api/views.py:453]:
```suggestion
class ContactUs(generics.CreateAPIView):
    serializer_class = ContactUsSerializer
    permission_classes = (permissions.AllowAny,)
```
There are some mistakes in that suggestion that wouldn't have the correct effect if committed directly

File: frontend/src/FullPageRoutes/AccountsPage.tsx (hunk 4 above)
Comment (jcarrete5) [reply frontend/src/FullPageRoutes/AccountsPage.tsx]: This is an error. I don't think it was caught because the event is declared with `any` type.

File: api/views.py [hunk 25]
```diff
@@ -416,3 +418,20 @@ class Accounts(generics.RetrieveAPIView):
...
+        serializer = self.serializer_class(data=request.data)
+        return response.Response(serializer.data, status=status.HTTP_201_CREATED)
+
+
+class Feedback(generics.CreateAPIView):
+    serializer_class = FeedbackSerializer
+
+    def post(self, request):
+        serializer = self.serializer_class(data=request.data)
+        return response.Response(serializer.data, status=status.HTTP_201_CREATED)
```
Comment (jcarrete5) [inline api/views.py]:
```suggestion
```
This method definition is redundant. This basic implementation is provided by default in `generics.CreateAPIView`

**Review by jcarrete5 — CHANGES_REQUESTED** [review]

File: api/views.py [hunk 26]
```diff
@@ -416,3 +418,20 @@ class Accounts(generics.RetrieveAPIView):
...
         return response.Response(serializer.data, status=status.HTTP_200_OK)
+
+
+class ContactUs(generics.CreateAPIView):
+    serializer_class = ContactUsSerializer
+    permission_clsses = permissions.AllowAny
+
+    def post(self, request):
+        serializer = self.serializer_class(data=request.data)
+        return response.Response(serializer.data, status=status.HTTP_201_CREATED)
```
Comment (jcarrete5) [inline api/views.py]:
```suggestion
```
This method is redundant for the same reason it is redundant in the feedback view

File: api/views.py (hunk 12 above)
Comment (sb3958) [reply api/views.py:453]: What will happen using this suggestion?

Comment (jcarrete5) [reply api/views.py:453]: The mistake appears to have already been committed. `permission_clsses` is misspelt. It should be `permission_classes` like in my revised suggestion. And the value should be a tuple. The comma is important after the `permissions.AllowAny` in the parentheses.

File: api/views.py (hunk 25 above)
Comment (sb3958) [reply api/views.py]: Since it is redundant, if I remove this what should I do then?

Comment (jcarrete5) [reply api/views.py]: You don't need to do anything except remove it. Removing the method should have the same effect as if it were there. Take a look at the `generics.CreateAPIView` source and you will find that the default implementation defines a post method that simply passes the data from the request to the serializer specified by the `serializer_class` class attribute and then saves the data.

File: api/models.py [hunk 27]
```diff
@@ -145,12 +145,12 @@ class ReportType(models.TextChoices):
...
-    message = models.CharField(max_length=2000, null=True)
+    message = models.TextField()
 
 
 class Feedback(models.Model):
     FEEDBACK_REASON = (("W", "Website"), ("Pp", "Penpal"), ("P", "Performance"))
 
-    reason = models.CharField(max_length=2, choices=FEEDBACK_REASON, null=True)
-    message = models.CharField(max_length=2000, blank=True)
+    reason = models.TextField(choices=FEEDBACK_REASON)
```
Comment (jcarrete5) [inline api/models.py]: This was fine as a `CharField` because it is only meant to store the two-character feedback reason. You might run into issues when you generate migrations if you change from `null=True` to `null=False` (defaults to `null=False`).
```suggestion
    reason = models.CharField(max_length=2, choices=FEEDBACK_REASON)
```

**Review by jcarrete5 — CHANGES_REQUESTED** [review]
Because changes have been made to some models, you will need to generate migrations again.

File: .vscode/settings.json [hunk 28]
```diff
@@ -1,3 +1,6 @@
 {
-    "python.pythonPath": "c:\\Users\\Sharelle\\Desktop\\Senior Design\\conteo\\venv\\Scripts\\python.exe"
+    
+    "workbench.editorAssociations": [
+    
+    ]
```
Comment (jcarrete5) [inline .vscode/settings.json]: You need to run `git rm -r --cached .vscode/` and commit to fully remove the `.vscode/` directory from the repository.

**Review by jcarrete5 — CHANGES_REQUESTED** [review]

**Review by elp49 — APPROVED** [review]

**Review by jcarrete5 — APPROVED** [review]

**Review by mneedleman — APPROVED** [review]
